pkg/reconciler/scheduling/namespace/namespace_controller.go

```diff
@@ -0,0 +1,380 @@
+	locationInformer schedulinginformers.LocationInformer,
+	workloadClusterInformer workloadinformers.WorkloadClusterInformer,
+	placementInformer schedulinginformers.PlacementInformer,
+) (*controller, error) {
+	queue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName)
+
+	c := &controller{
+		queue: queue,
+		enqueueAfter: func(clusterName logicalcluster.Name, ns *corev1.Namespace, duration time.Duration) {
+			key := clusters.ToClusterAwareKey(clusterName, ns.Name)
+			queue.AddAfter(key, duration)
+		},
+
+		kubeClusterClient: kubeClusterClient,
+		kcpClusterClient:  kcpClusterClient,
+
+		namespaceLister:  namespaceInformer.Lister(),
+		namespaceIndexer: namespaceInformer.Informer().GetIndexer(),
+
+		locationLister:  locationInformer.Lister(),
+		locationIndexer: locationInformer.Informer().GetIndexer(),
+
+		workloadClusterLister:  workloadClusterInformer.Lister(),
+		workloadClusterIndexer: workloadClusterInformer.Informer().GetIndexer(),
+
+		placmentLister:   placementInformer.Lister(),
+		placementIndexer: placementInformer.Informer().GetIndexer(),
+	}
+
+	if err := locationInformer.Informer().AddIndexers(cache.Indexers{
+		byWorkspace: indexByWorksapce,
+	}); err != nil {
+		return nil, err
+	}
+
+	if err := workloadClusterInformer.Informer().AddIndexers(cache.Indexers{
+		byWorkspace: indexByWorksapce,
+	}); err != nil {
+		return nil, err
+	}
+
+	if err := namespaceInformer.Informer().AddIndexers(cache.Indexers{
+		byWorkspace: indexByWorksapce,
+	}); err != nil {
+		return nil, err
+	}
+
+	if err := placementInformer.Informer().AddIndexers(cache.Indexers{
+		byWorkspace:         indexByWorksapce,
+		byLocationWorkspace: indexByLoactionWorkspace,
+	}); err != nil {
+		return nil, err
+	}
+
+	// namespaceBlocklist holds a set of namespaces that should never be synced from kcp to physical clusters.
+	var namespaceBlocklist = sets.NewString("kube-system", "kube-public", "kube-node-lease")
+	namespaceInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			switch ns := obj.(type) {
+			case *corev1.Namespace:
+				return !namespaceBlocklist.Has(ns.Name)
+			case cache.DeletedFinalStateUnknown:
+				return true
+			default:
+				return false
+			}
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    func(obj interface{}) { c.enqueueNamespace(obj) },
+			UpdateFunc: func(_, obj interface{}) { c.enqueueNamespace(obj) },
+			DeleteFunc: func(obj interface{}) { c.enqueueNamespace(obj) },
+		},
+	})
+
+	locationInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) { c.enqueueLocation(obj) },
+			UpdateFunc: func(old, obj interface{}) {
+				oldLoc := old.(*schedulingv1alpha1.Location)
+				newLoc := obj.(*schedulingv1alpha1.Location)
+				if !reflect.DeepEqual(oldLoc.Spec, newLoc.Spec) || !reflect.DeepEqual(oldLoc.ObjectMeta, newLoc.ObjectMeta) {
```

Comment:
I think we have to compare labels and annotations here. The RV is in ObjectMeta, which means this comparison is always true if status changes.